### DISCUSSION TITLE
Can I initial my postgresql provider with my docker

### DISCUSSION THREAD

**Discussion by JanneHarju** [drakkan/sftpgo#546] [Q&A] [answered]
Hello, i'm trying to get sftpgo working with docker, because my initial goal is to get it work in azure container with keyVault and blob storage. At first I'm getting familiar with sftpgo in my local machine. So as header asks, can I initial my database using only docker or am I forced to install cli? I'm using windows 10.

**Comment by drakkan** [comment, marked as answer]
Hi,

I'm not sure to understand you question, you have several options:

1) you can install SFTPGo directly on Windows, download link [here](https://github.com/drakkan/sftpgo/releases/download/v2.1.2/sftpgo_v2.1.2_windows_x86_64.exe). It will run as service using SQLite provider. It just works, you can connect to the web ui and start using it. Please take a look at the getting started guide [here](https://github.com/drakkan/sftpgo/blob/main/docs/howto/getting-started.md)
2) you can configure your windows service to connect to a Postgres instance editing the sftpgo.json configuration file and then restarting the SFTPGo service. Postgres can run natively on windows or on docker
3) you can install both sftpgo and postgres in docker. I'm not familiar with Docker desktop for windows. Please take a look [here](https://github.com/drakkan/sftpgo/tree/main/docker) to understand the docker commands required to configure SFTPGo. You can configure postgres access via env [vars](https://github.com/drakkan/sftpgo/blob/main/docs/full-configuration.md#environment-variables)

If you want to change the default configuration, for example to connect to Postgres, it is not difficult but it also depends on your background, you probably have to do some steps using a cli and / or a text editor.

I also offer pre-configured Amazon Machine images, postgres is available, obviously this is not free (but using them you will help the project)

https://aws.amazon.com/marketplace/seller-profile?id=6e849ab8-70a6-47de-9a43-13c3fa849335

thanks

**Reply by JanneHarju** [reply]
Thank you for very quick and comprehensive response.

So we are using AWS but moving to Azure so AWS is not option anymore to us.

I have read docker manual as well as configuration part of sftpgo. I got docker running without postgreSQL with default command in your documentation. I combined postgre tutorial with docker and managed to write this command. I'm running postgreSQL also with docker.
`docker run --name project-sftpgo -p 8080:8080 -p 2022:2022 -e SFTPGO_DATA_PROVIDER__DRIVER='postgresql' -e SFTPGO_DATA_PROVIDER__NAME='sftpgo.db' -e SFTPGO_DATA_PROVIDER__HOST='host' -e SFTPGO_DATA_PROVIDER__PORT=5432 -e SFTPGO_DATA_PROVIDER__USERNAME='sftpgo' -e SFTPGO_DATA_PROVIDER__PASSWORD='sftpgo_pg_pwd' -e SFTPGO_DATA_PROVIDER__SSLMODE=1 -d "drakkan/sftpgo:latest"`
before I run that command I create user and database with these suggested commands.
`create user "sftpgo" with encrypted password 'sftpgo_pg_pwd';
create database "sftpgo.db";
grant all privileges on database "sftpgo.db" to "sftpgo";`

After starting docker I follow logs inside container. It says that it get connected to database but then it follows with error.
`{"level":"warn","time":"2021-09-20T16:38:41.612","sender":"dataprovider_postgresql","message":"error preparing database query \"SELECT version from schema_version LIMIT 1\": dial tcp: i/o timeout"}

{"level":"warn","time":"2021-09-20T16:38:51.613","sender":"dataprovider_postgresql","message":"Unable to initialize data provider: dial tcp: i/o timeout"}

{"level":"error","time":"2021-09-20T16:38:51.613","sender":"service","message":"error initializing data provider: dial tcp: i/o timeout"}`

Now I start to think that maybe there is still some connection problem. At first it said connection actively refused to connect, but now it says timeout. I will double check connection parameters.

**Reply by JanneHarju** [reply]
I forget to ask that is docker container trying to run initialprovider after docker container has bee started? Or should initialprovider command be runned before starting docker container (with cli command)?

**Reply by JanneHarju** [reply]
Sorry for spamming. I started to wonder that maybe I must make Docker compose file and manage connection through it. It might be hard(impossible) to get docker container to connect to other docker container without compose. Because postgresql is not running in host it is running in docker. I will try that tomorrow.

**Reply by drakkan** [reply]
Yes, docker compose should help

**Reply by JanneHarju** [reply]
I managed it to work. Thank you for help. Now trying to learn how configure it with Azure blob storage. It seems to be quite straight forward with UI. After that I will check how I can use Azure Key Vault. In documentation it says that it is still experimental. Do you have some guide for configuring it? I didn't find manual.

**Reply by JanneHarju** [reply]
To answer to myself https://github.com/sftpgo/sftpgo-plugin-kms#azure-keyvault This seems to be enough.

**Reply by drakkan** [reply]
> To answer to myself https://github.com/sftpgo/sftpgo-plugin-kms#azure-keyvault This seems to be enough.

this should work, please note that plugins are supported only in `edge*` tags for now so you have to use one of these docker images

**Reply by JanneHarju** [reply]
start to wonder what keys are going to KeyVault? For example in Azure blob storage? Account Key, SAS URL? And for users Public key? So what Kay Vault is used for in sftpgo?

**Reply by drakkan** [reply]
> start to wonder what keys are going to KeyVault? For example in Azure blob storage? Account Key, SAS URL? And for users Public key? So what Kay Vault is used for in sftpgo?

Please take a look at the docs [here](https://github.com/drakkan/sftpgo/blob/main/docs/kms.md) and at the schema definition [here](https://sftpgo.stoplight.io/docs/sftpgo/c2NoOjc1MDgzNDQ-azure-blob-fs-config). 


For azblob config  you are right, AccountKey and SAS URL  can be stored in key vault.


If you don't configure an external KMS they will be stored using the local provider. This is secure if you define a master key and keep it secret (you can use a docker secret or something similar).

Public keys can be shared securely, there is no need to encrypt them
